brut.apktool/apktool-lib/src/main/java/brut/androlib/res/decoder/StringBlock.java

```diff
@@ -98,10 +109,154 @@ public String getString(int index) {
+            }
+            throw new IllegalStateException();
+        }
+
+        private enum Type {
+            OPEN,
+            CLOSE
+        }
+    }
+
+    private static class Span {
+        private String tag;
+        private int firstChar, lastChar;
+
+        Span(String val, int firstIndex, int lastIndex) {
+            this.tag = val;
+            this.firstChar = firstIndex;
+            this.lastChar = lastIndex;
+        }
+
+        String getTag() {
+            return tag;
+        }
+
+        int getFirstChar() {
+            return firstChar;
+        }
+
+        int getLastChar() {
+            return lastChar;
+        }
+    }
+
+    private static class StyledString {
+        String val;
+        int[] styles;
+
+        StyledString(String raw, int[] stylesArr) {
+            this.val = raw;
+            this.styles = stylesArr;
+        }
+
+        String getValue() {
+            return val;
+        }
+
+        List<Span> getSpanList(StringBlock stringBlock) {
+            ArrayList<Span> spanList = new ArrayList<>();
+            for (int i = 0; i != styles.length; i += 3) {
+                spanList.add(new Span(stringBlock.getString(styles[i]), styles[i + 1], styles[i + 2]));
+            }
+            return spanList;
+        }
+    }
+
+    String processStyledString(StyledString styledString) {
+
+        ArrayList<Tag> sortedTagsList = new ArrayList<>();
+
+        styledString.getSpanList(this).stream()
+            .flatMap(
+                span ->
+                    Stream.of(
+                        // "+ 1" because the last char is included.
+                        new Tag(
+                            span.getTag(), Tag.Type.OPEN, span.getFirstChar(), span.getLastChar() + 1),
+                        // "+ 1" because the last char is included.
+                        new Tag(
+                            span.getTag(),
+                            Tag.Type.CLOSE,
+                            span.getLastChar() + 1,
+                            span.getFirstChar())))
+            // So we can edit the string in place, we need to start from the end.
+            .sorted(naturalOrder())
+            .forEach(tag -> sortedTagsList.add(tag));
+
+        String raw = styledString.getValue();
+        StringBuilder string = new StringBuilder(raw.length() + 32);
+        int lastIndex = 0;
+        for (Tag tag : sortedTagsList) {
+            string.append(ResXmlEncoders.escapeXmlChars(raw.substring(lastIndex, tag.position)));
+            string.append(tag.toString());
+            lastIndex = tag.position;
+        }
+        string.append(ResXmlEncoders.escapeXmlChars(raw.substring(lastIndex)));
+
+        return string.toString();
+    }
+
+    /**
+     * Not yet implemented.
+     *
+     * Returns string with style information (if any).
+     */
+    public CharSequence get(int index) {
+        return getString(index);
+    }
+
```

Comment:
Was this function just part of the port? I can't see where its used.

Reply:
my bad. ince i based my original code on 2.5.0 this was copied from there. can be removed then.